ArC: implement CDI 4.1

Ladicek opened the pull request.

Ladicek committed:
ArC: update CDI to 4.1.0, Interceptors to 2.2.0 and Common Annotations to 3.0.0

Ladicek committed:
ArC: implement the CreationalContexts TCK SPI

Ladicek committed:
ArC: implement the Contextuals TCK SPI

Ladicek committed:
ArC: prefer InvocationContext.getInterceptorBindings()

Ladicek committed:
ArC: implement BeanContainer.getContexts()

Ladicek committed:
ArC: fix ambiguity resolution in case of producers with priority

A producer may have priority as defined by CDI 4.1 (or obtained from
a stereotype, as permitted but not really specified in CDI 4.0),
but does not necessarily have to be an alternative. During ambiguity
resolution, if such non-alternative producer with priority is declared
on an alternative bean, we must first consider the priority declared
on the producer, even if the producer is not itself an alternative.

To do that, this commit changes the ambiguity resolution implementation
to use `BeanInfo.getPririty()` instead of `getAlternativePriority()`.
The difference is that `getAlternativePriority()` returns `null` when
the bean itself is not an alternative. Given that the method was
previously only used during ambiguity resolution, and only after
non-alternative beans (and producers on non-alternative beans)
were eliminated, this is correct.

In fact, this commit also marks `BeanInfo.getAlternativePriority()`
as deprecated for removal. For all practical purposes, `getPriority()`
is a better choice; it is usually meaningless to consider a bean
priority only when the bean is an alternative, and pretend that
the bean has no priority when it is not an alternative.

Ladicek committed:
ArC: support registering multiple custom contexts for single scope

Ladicek committed:
ArC: add support for method invokers

Ladicek committed:
ArC: implement BeanContainer.isMatchingBean() and isMatchingEvent()

Ladicek committed:
ArC: improve validation of built-in beans injection

Ladicek committed:
ArC: centralize code generation for wrapping InjectableReferenceProvider with CurrentInjectionPointProvider

Ladicek committed:
ArC: fix CustomNormalScopeTest, we can use BeanContainer.getContexts() now

Ladicek committed:
ArC: remove extra test/java17 directory, now that Java 17 is the minimum

Ladicek committed:
ArC: update documentation to CDI 4.1

Ladicek commented:
Each commit is best reviewed separately. They are all fairly small, except of the one that adds support for method invokers.

gsmet commented:
Not a review but nice work :).

Ladicek commented:
Rebased.

mkouba commented on independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/BeanManagerImpl.java:
```diff
@@ -226,6 +228,11 @@ public Context getContext(Class<? extends Annotation> scopeType) {
 
+    @Override
+    public Collection<Context> getContexts(Class<? extends Annotation> scopeType) {
+        return Collections.unmodifiableList(Arc.container().getContexts(scopeType));
```
I believe that `Arc.container().getContexts(scopeType)` already returns an immutable list, although it's not mentioned in the javadoc. So maybe just update the javadoc and skip the `Collections.unmodifiableList()` here.

Ladicek replied:
OK, I didn't dive too much into this previously, but you're right -- `io.quarkus.arc.impl.Contexts` always returns immutable lists. I'll change this.

Ladicek replied:
Done.

mkouba commented on independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanGenerator.java:
```diff
@@ -2228,9 +2229,10 @@ private String getProxyTypeName(BeanInfo bean, String baseName) {
-    private ResultHandle wrapCurrentInjectionPoint(BeanInfo bean,
-            MethodCreator constructor, InjectionPointInfo injectionPoint, int paramIdx, ResultHandle tccl,
-            ReflectionRegistration reflectionRegistration) {
+    public static ResultHandle wrapCurrentInjectionPoint(BeanInfo bean, MethodCreator constructor,
```
Just nitpicking, but I prefer package-private modifier where possible...

Ladicek replied:
Hmm, I wonder why I made this public. There's been no need for that... Let me fix.

mkouba reviewed (approved):
@Ladicek do you plan to add support for Quarkus build items used to define method invokers?

Ladicek commented:
I don't think that's necessary. Method invokers can be registered in the bean registration phase, for which a build item already exists. (Aside: maybe this should be better documented?) I have a few examples (`@Scheduled`, `@ConsumeEvent`, Reactive Messaging) that I hope to submit after this is merged and I didn't need any new build items.

mkouba commented:
> I don't think that's necessary. Method invokers can be registered in the bean registration phase, for which a build item already exists. (Aside: maybe this should be better documented?) I have a few examples (`@Scheduled`, `@ConsumeEvent`, Reactive Messaging) that I hope to submit after this is merged and I didn't need any new build items.

With `BeanRegistrar#getInvokerFactory()` you mean? Yes, that's hidden very well :-). OTOH we also have `SyntheticBeanBuildItem` and `BeanRegistrar.RegistrationContext#configure()` but I'm not sure if a dedicated build item  would bring any benefit here...

Ladicek commented on independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/Types.java:
```diff
@@ -215,6 +215,25 @@ static boolean containsTypeVariable(Type type) {
+            ParameterizedType parameterizedType = asParameterizedType(type);
+            for (Type typeArgument : parameterizedType.getActualTypeArguments()) {
+                // the 2nd condition is a bit weird, because the spec doesn't say
+                // anything about illegal type arguments, but Weld has it...
```
Hmm, I have no idea why I didn't copy that. I'll fix it.

Ladicek replied:
Done.

Ladicek commented on independent-projects/arc/runtime/src/main/java/io/quarkus/arc/ArcInvocationContext.java:
```diff
@@ -17,32 +17,25 @@ public interface ArcInvocationContext extends InvocationContext {
      */
     String KEY_INTERCEPTOR_BINDINGS = "io.quarkus.arc.interceptorBindings";
```
Fair enough. I didn't originally deprecate this, because it's used on many places in ArC, but I can do that.

Ladicek replied:
Done.

Ladicek commented:
> With `BeanRegistrar#getInvokerFactory()` you mean? Yes, that's hidden very well :-)

Indeed that's what I meant :-) Yeah, I agree. I should probably write some documentation into the CDI integration guide...

manovotn reviewed (approved):
With the review of invoker methods, I admit I mostly went after how the generated code looks like and what test coverage we have (which is very nice BTW).

The good thing is that we did no changes or removals in CDI 4.0 that would break other projects (such as SR which are built against 4.0) so we should be able to get this in smoothly.

Looks good!

Ladicek commented:
The TCK failures are probably caused by an old TestNG version. Will take a look.

Ladicek committed:
ArC: fix running the CDI TCK

Ladicek committed:
Update TestNG to 7.8.0, a version used in the MicroProfile parent POM

Ladicek commented:
Rebased and added 2 more commits.

This will fail on RESTEasy Reactive TCK, which uses `@ManagedBean` in a few classes, which is gone from Annotations 3.0. Not sure what to do about that... probably keep Annotations on 2.1.1 until JAX-RS releases and RESTEasy Reactive updates?

gastaldi commented:
I'd prefer going Annotations 3.0. Maybe include the RESTEasy Reactive and JAX-RS updates in this same PR too?

Ladicek commented:
Here's a PR to remove the test for `@ManagedBean` from the RESTEasy Reactive TCK: https://github.com/quarkusio/resteasy-reactive-testsuite/pull/7 Will update this PR after that one is merged.

Ladicek committed:
Update the RESTEasy Reactive TCK

The new version doesn't contain the `@ManagedBean` annotation,
which was removed in Jakarta Annotations 3.0.

Ladicek commented:
Rebased and updated the RESTEasy Reactive TCK.

gsmet commented:
@mkouba @manovotn all good for you? CI looks happy. I let you merge if you are :).
